src/test/operators/join_test_runner.cpp

```diff
@@ -0,0 +1,527 @@
+                                            configuration.radix_bits);
+    } else {
+      return std::make_shared<JoinOperator>(left, right, mode, primary_predicate, secondary_predicates);
+    }
+  }
+};
+// Order of columns in the input tables
+const std::unordered_map<DataType, size_t> data_type_order = {
+    {DataType::Int, 0u}, {DataType::Float, 1u}, {DataType::Double, 2u}, {DataType::Long, 3u}, {DataType::String, 4u},
+};
+
+}  // namespace
+
+namespace opossum {
+
+class JoinTestRunner : public BaseTestWithParam<JoinTestConfiguration> {
+ public:
+  template <typename JoinOperator>
+  static std::vector<JoinTestConfiguration> create_configurations() {
+    auto configurations = std::vector<JoinTestConfiguration>{};
+
+    /**
+     * For each `all_*` set, the first element is the default - which is chosen to be sensible. Thus the elements are
+     * not necessarily sorted.
+     */
+    auto all_data_types = std::vector<DataType>{};
+    hana::for_each(data_type_pairs, [&](auto pair) {
+      const DataType d = hana::first(pair);
+      all_data_types.emplace_back(d);
+    });
+
+    const auto all_predicate_conditions = std::vector{
+        PredicateCondition::Equals,      PredicateCondition::NotEquals,
+        PredicateCondition::GreaterThan, PredicateCondition::GreaterThanEquals,
+        PredicateCondition::LessThan,    PredicateCondition::LessThanEquals,
+    };
+    const auto all_join_modes = std::vector{JoinMode::Inner,         JoinMode::Left, JoinMode::Right,
+                                            JoinMode::FullOuter,     JoinMode::Semi, JoinMode::AntiNullAsFalse,
+                                            JoinMode::AntiNullAsTrue};
+    const auto all_table_sizes = std::vector{10u, 15u, 0u};
+    const auto all_left_nulls = std::vector{true, false};
+    const auto all_right_nulls = std::vector{true, false};
+    const auto all_chunk_sizes = std::vector{10u, 3u, 1u};
+    const auto all_secondary_predicate_sets = std::vector<std::vector<OperatorJoinPredicate>>{
+        {},
+        {{{ColumnID{0}, ColumnID{0}}, PredicateCondition::LessThan}},
+        {{{ColumnID{0}, ColumnID{0}}, PredicateCondition::GreaterThanEquals}},
+        {{{ColumnID{0}, ColumnID{0}}, PredicateCondition::NotEquals}}};
+
+    const auto all_swap_input_sides = std::vector{false, true};
+    const auto all_input_table_types =
+        std::vector{InputTableType::Data, InputTableType::IndividualPosLists, InputTableType::SharedPosList};
+
+    // clang-format off
+    JoinTestConfiguration default_configuration{
+      InputTableConfiguration{
+        InputSide::Left, all_chunk_sizes.front(), all_table_sizes.front(), all_input_table_types.front()},
+      InputTableConfiguration{
+        InputSide::Right, all_chunk_sizes.front(), all_table_sizes.front(), all_input_table_types.front()},
+      JoinMode::Inner,
+      DataType::Int,
+      DataType::Int,
+      true,
+      true,
```

Comment:
Enable c++20 and bring in designated initializers?

Reply:
Given how red/orange clang still is, I'd say no: https://clang.llvm.org/cxx_status.html